ubuntu-advantage-client/ec2_cleanup.py

```diff
@@ -19,62 +19,125 @@ def parse_args():
         help=("Tag used to filter cloud resources for deletion. "
               "Default: {}".format(CI_DEFAULT_TAG))
     )
+    parser.add_argument(
+        "-o", "--older-than", dest="older_than", action="store",
+        help=("Tag used to filter cloud resources for deletion. "
+              "Format: MM/DD/YY [HH:MM:SS].)"
+    )
     return parser.parse_args()
 
 
-def clean_ec2(tag):
+def get_tag_prefix(older_than):
+    if not older_than:
+       return ""
+    print('Match only resources created before %s' % older_than)
+    try:
+        time = datetime.datetime.strptime(older_than, "%m/%d/%y %H:%M:%S")
+        return time.strftime("uaclient-ci-%m%d-%H%M%S")
+    except ValueError:
+        time = datetime.datetime.strptime(older_than, "%m/%d/%y")
+        return time.strftime("uaclient-ci-%m%d")
+
+
+def is_resource_stale(resource, tag_time_prefix):
+    """Return True if the given resource is older than the tag_time_prefix.
+
+    If no Name tag present, assume stale and return True
+    """
+    for tag in resource.tags:
+        if tag["Key"] != "Name":
+            continue
+        if tag["Value"] >= tag_time_prefix:
+            return False
+    return True
+
+
+def clean_ec2(tag_prefix, older_than=None):
     """Clean up all running EC2 instances, VPCs, and storage."""
     client = boto3.client('ec2')
     resource = boto3.resource('ec2')
 
-    print('# searching for vpcs matching tag {}'.format(tag))
-    tag_filter = [{'Name': 'tag:Name', 'Values': [tag]}]
-    for vpc in list(resource.vpcs.filter(Filters=tag_filter)):
+    tag_time_prefix = get_tag_prefix(older_than)
+    print('# searching for vpcs matching tag {}'.format(
+        "uaclient-integration")
+    )
+    tag_filter = [{'Name': 'tag:Name', 'Values': [tag_prefix]}]
+    vpc_filter = [{'Name': 'tag:Name', 'Values': ["uaclient-integration"]}]
+    for vpc in list(resource.vpcs.filter(Filters=vpc_filter)):
         print('cleaning up vpc %s' % vpc.id)
+        wait_instances = []
+        skipped_instances = []
         for instance in vpc.instances.all():
+            if tag_time_prefix:
+                if not is_resource_stale(instance, tag_time_prefix):
+                    skipped_instances.append(instance)
             print('terminating instance %s' % instance.id)
             instance.terminate()
-            instance.wait_until_terminated()
+            wait_instances.append(instance)
+
+        for inst in skipped_instances:
+            print(
+                "left instance %s running; newer than %s" % (
+                    inst.id, older_than
+                )
+            )
+        for inst in wait_instances:
+            print("waiting on %s" % inst.id)
+            inst.wait_until_terminated()
 
         for security_group in vpc.security_groups.filter(Filters=tag_filter):
+            if not should_delete_resource(security_group, tag_time_prefix):
```

Comment:
Maybe I am missing something, but I could not find the definition of `should_delete_resource`